Add config to exempt local vars from ExplicitTypeInterfaceRule

rounak opened the pull request:
This PR adds a new `excluded` config to this existing rule, to exempt certain kinds of variables from having explicit types. Values can be `local`, `static`, `class` & `instance`.

Would love to add test coverage for this, but didn't know if I should write a new test in `RuleTests` or do it some other way.

marcelofabri commented on CHANGELOG.md:
```diff
@@ -6,7 +6,10 @@
 
-* None.
+* Add a new `excluded` config parameter to the `explicit_type_interface` rule
+  to exempt certain types of variables from the rule. 
```
you should leave two trailing whitespaces (currently just one)

marcelofabri commented on Source/SwiftLintFramework/Rules/RuleConfigurations/ExplicitTypeInterfaceConfiguration.swift:
```diff
@@ -0,0 +1,58 @@
+public struct ExplicitTypeInterfaceConfiguration: RuleConfiguration, Equatable {
+    public var severityConfiguration = SeverityConfiguration(.warning)
+
+    var allowedKinds: Set<SwiftDeclarationKind> = [.varInstance,
```
this should be `public`

marcelofabri commented on Source/SwiftLintFramework/Rules/RuleConfigurations/ExplicitTypeInterfaceConfiguration.swift:
```diff
@@ -0,0 +1,58 @@
+                                                   .varClass]
+
+    public var consoleDescription: String {
+        return severityConfiguration.consoleDescription + ", allowed kinds: \(allowedKinds)"
```
can we print the allowedKinds in the same short style that is expect to be in the configuration file?

marcelofabri commented:
> Would love to add test coverage for this, but didn't know if I should write a new test in RuleTests or do it some other way.

You can create a new file `ExplicitTypeInterfaceConfigurationTests` (take a look at `CyclomaticComplexityConfigurationTests` for example) and another one to test the rule itself (`ExplicitTypeInterfaceRuleTests`).

When you create `ExplicitTypeInterfaceRuleTests`, you should move the existing test in `RulesTests` to the new file, along with new tests that test different configurations. You can take a look at existing tests to see how it's usually done (`ColonRuleTests` would be an example).

marcelofabri commented on Source/SwiftLintFramework/Rules/RuleConfigurations/ExplicitTypeInterfaceConfiguration.swift:
```diff
@@ -27,27 +27,37 @@ private extension SwiftDeclarationKind {
     public var consoleDescription: String {
-        return severityConfiguration.consoleDescription + ", allowed kinds: \(allowedKinds)"
+        return severityConfiguration.consoleDescription +
+        ", excluded: [\(ExplicitTypeInterfaceConfiguration.variableKinds.subtracting(allowedKinds))]"
```
My comment here was more about how the types will be spelled out.

By just printing `SwiftDeclarationKind`, you'll get something like `source.lang.swift.decl.var.class`, but when configuring the rule you're expected to use just `class`.

rounak replied:
Ah I see, let me do that

rounak commented:
Thanks for the review @marcelofabri.

Added two new test files as you said, one for config and one for the rule itself. The only issue is since this local variable stuff depends on SourceKit, we can't add concrete test cases (it warns for local vars) until everyone's running Xcode 9.3.

As for the consoleDescription, I didn't quite understand:

> the same short style that is expect to be in the configuration file

I added square brackets around the set, is that what you meant?

marcelofabri commented:
> The only issue is since this local variable stuff depends on SourceKit, we can't add concrete test cases (it warns for local vars) until everyone's running Xcode 9.3.

Can you add a test excluding another kind (e.g. class variables)?

rounak commented:
@marcelofabri Done, let me know if it looks okay now.

marcelofabri commented on Tests/SwiftLintFrameworkTests/ExplicitTypeInterfaceTests.swift:
```diff
@@ -0,0 +1,47 @@
+import SwiftLintFramework
+import XCTest
+
+class ExplicitTypeInterfaceTests: XCTestCase {
```
Can you rename this class (and file) to `ExplicitTypeInterfaceRuleTests`?

marcelofabri commented:
@rounak This looks good! I have one minor nitpick that I added as a comment.

Can you also rebase against master now that I merged #2068? That should fix the CI failure.

Thanks!
